--- a/src/unix.rs
+++ b/src/unix.rs
@@ -139,7 +139,8 @@
           target_os = "netbsd",
           target_os = "dragonfly",
           target_os = "solaris",
-          target_os = "haiku"))]
+          target_os = "haiku",
+          target_os = "zos"))]
 pub fn allocate(file: &File, len: u64) -> Result<()> {
     // No file allocation API available, just set the length if necessary.
     if len > try!(file.metadata()).len() as u64 {
