--- a/build.rs
+++ b/build.rs
@@ -138,7 +138,10 @@
 
     // Either include ASM files, or disable ASM entirely.
     // Also disable it on windows, apparently it doesn't do well with these .S files at the moment.
-    if cfg!(feature = "no_asm") || std::env::var("CARGO_CFG_WINDOWS").is_ok() {
+    if cfg!(feature = "no_asm")
+        || std::env::var("CARGO_CFG_WINDOWS").is_ok()
+        || std::env::var("CARGO_CFG_TARGET_OS").as_deref() == Ok("zos")
+    {
         config.define("ZSTD_DISABLE_ASM", Some(""));
     } else {
         config.file("zstd/lib/decompress/huf_decompress_amd64.S");
