--- a/src/backend/libc/io/syscalls.rs
+++ b/src/backend/libc/io/syscalls.rs
@@ -101,6 +101,7 @@
     target_os = "redox",
     target_os = "solaris",
     target_os = "vita",
+    target_os = "zos",
 )))]
 pub(crate) fn preadv(
     fd: BorrowedFd<'_>,
@@ -132,6 +133,7 @@
     target_os = "redox",
     target_os = "solaris",
     target_os = "vita",
+    target_os = "zos",
 )))]
 pub(crate) fn pwritev(fd: BorrowedFd<'_>, bufs: &[IoSlice<'_>], offset: u64) -> io::Result<usize> {
     // Silently cast; we'll get `EINVAL` if the value is negative.
@@ -267,6 +269,8 @@
 #[cfg(not(any(
     apple,
     target_os = "aix",
+
+    target_os = "zos",
     target_os = "android",
     target_os = "dragonfly",
     target_os = "espidf",
@@ -293,6 +297,7 @@
     target_os = "dragonfly",
     target_os = "haiku",
     target_os = "redox",
+    target_os = "zos",
 ))]
 pub(crate) fn dup3(fd: BorrowedFd<'_>, new: &mut OwnedFd, _flags: DupFlags) -> io::Result<()> {
     // Android 5.0 has `dup3`, but libc doesn't have bindings. Emulate it
--- a/src/backend/libc/fs/syscalls.rs
+++ b/src/backend/libc/fs/syscalls.rs
@@ -1283,6 +1283,7 @@
     target_os = "redox",
     target_os = "solaris",
     target_os = "vita",
+    target_os = "zos",
 )))]
 pub(crate) fn fadvise(
     fd: BorrowedFd<'_>,
@@ -1638,7 +1639,7 @@
     }
 }
 
-#[cfg(not(any(target_os = "espidf", target_os = "horizon", target_os = "vita")))]
+#[cfg(not(any(target_os = "espidf", target_os = "horizon", target_os = "vita", target_os = "zos")))]
 pub(crate) fn futimens(fd: BorrowedFd<'_>, times: &Timestamps) -> io::Result<()> {
     // Old 32-bit version: libc has `futimens` but it is not y2038 safe by
     // default. But there may be a `__futimens64` we can use.
--- a/src/backend/libc/mm/syscalls.rs
+++ b/src/backend/libc/mm/syscalls.rs
@@ -177,6 +177,7 @@
 ///
 /// `mlock` operates on raw pointers and may round out to the nearest page
 /// boundaries.
+#[cfg(not(target_os = "zos"))]
 #[inline]
 pub(crate) unsafe fn mlock(addr: *mut c::c_void, length: usize) -> io::Result<()> {
     ret(c::mlock(addr, length))
@@ -208,6 +209,7 @@
 ///
 /// `munlock` operates on raw pointers and may round out to the nearest page
 /// boundaries.
+#[cfg(not(target_os = "zos"))]
 #[inline]
 pub(crate) unsafe fn munlock(addr: *mut c::c_void, length: usize) -> io::Result<()> {
     ret(c::munlock(addr, length))
--- a/src/mm/mmap.rs
+++ b/src/mm/mmap.rs
@@ -308,6 +308,7 @@
 /// [DragonFly BSD]: https://man.dragonflybsd.org/?command=mlock&section=2
 /// [illumos]: https://illumos.org/man/3C/mlock
 /// [glibc]: https://sourceware.org/glibc/manual/latest/html_node/Page-Lock-Functions.html#index-mlock
+#[cfg(not(target_os = "zos"))]
 #[inline]
 pub unsafe fn mlock(ptr: *mut c_void, len: usize) -> io::Result<()> {
     backend::mm::syscalls::mlock(ptr, len)
@@ -372,6 +373,7 @@
 /// [DragonFly BSD]: https://man.dragonflybsd.org/?command=munlock&section=2
 /// [illumos]: https://illumos.org/man/3C/munlock
 /// [glibc]: https://sourceware.org/glibc/manual/latest/html_node/Page-Lock-Functions.html#index-munlock
+#[cfg(not(target_os = "zos"))]
 #[inline]
 pub unsafe fn munlock(ptr: *mut c_void, len: usize) -> io::Result<()> {
     backend::mm::syscalls::munlock(ptr, len)
--- a/src/io/read_write.rs
+++ b/src/io/read_write.rs
@@ -222,6 +222,7 @@
     target_os = "redox",
     target_os = "solaris",
     target_os = "vita",
+    target_os = "zos",
 )))]
 #[inline]
 pub fn preadv<Fd: AsFd>(fd: Fd, bufs: &mut [IoSliceMut<'_>], offset: u64) -> io::Result<usize> {
@@ -261,6 +262,7 @@
     target_os = "redox",
     target_os = "solaris",
     target_os = "vita",
+    target_os = "zos",
 )))]
 #[inline]
 pub fn pwritev<Fd: AsFd>(fd: Fd, bufs: &[IoSlice<'_>], offset: u64) -> io::Result<usize> {
--- a/src/fs/fadvise.rs
+++ b/src/fs/fadvise.rs
@@ -18,6 +18,18 @@
 /// [FreeBSD]: https://man.freebsd.org/cgi/man.cgi?query=posix_fadvise&sektion=2
 #[inline]
 #[doc(alias = "posix_fadvise")]
+#[cfg(not(any(
+    apple,
+    netbsdlike,
+    target_os = "dragonfly",
+    target_os = "espidf",
+    target_os = "haiku",
+    target_os = "horizon",
+    target_os = "redox",
+    target_os = "solaris",
+    target_os = "vita",
+    target_os = "zos",
+)))]
 pub fn fadvise<Fd: AsFd>(
     fd: Fd,
     offset: u64,
--- a/src/fs/fd.rs
+++ b/src/fs/fd.rs
@@ -210,7 +210,7 @@
 ///
 /// [POSIX]: https://pubs.opengroup.org/onlinepubs/9799919799/functions/futimens.html
 /// [Linux]: https://man7.org/linux/man-pages/man2/utimensat.2.html
-#[cfg(not(any(target_os = "espidf", target_os = "horizon", target_os = "vita")))]
+#[cfg(not(any(target_os = "espidf", target_os = "horizon", target_os = "vita", target_os = "zos")))]
 #[inline]
 pub fn futimens<Fd: AsFd>(fd: Fd, times: &Timestamps) -> io::Result<()> {
     backend::fs::syscalls::futimens(fd.as_fd(), times)
--- a/src/fs/mod.rs
+++ b/src/fs/mod.rs
@@ -82,6 +82,7 @@
     target_os = "redox",
     target_os = "solaris",
     target_os = "vita",
+    target_os = "zos",
 )))]
 pub use fadvise::fadvise;
 pub use fcntl::*;
