diff -ruN '--exclude=*.orig' '--exclude=Cargo.lock' /home/itodorov/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/getrandom-0.4.3/Cargo.toml /tmp/getrandom-04-zos/Cargo.toml
--- /home/itodorov/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/getrandom-0.4.3/Cargo.toml	2006-07-23 21:21:28.000000000 -0400
+++ /tmp/getrandom-04-zos/Cargo.toml	2026-08-11 10:46:00.160166848 -0400
@@ -108,7 +108,7 @@
 version = "0.2.154"
 default-features = false
 
-[target.'cfg(any(target_os = "haiku", target_os = "redox", target_os = "nto", target_os = "aix"))'.dependencies.libc]
+[target.'cfg(any(target_os = "haiku", target_os = "redox", target_os = "nto", target_os = "aix", target_os = "zos"))'.dependencies.libc]
 version = "0.2.154"
 default-features = false
 
diff -ruN '--exclude=*.orig' '--exclude=Cargo.lock' /home/itodorov/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/getrandom-0.4.3/src/backends.rs /tmp/getrandom-04-zos/src/backends.rs
--- /home/itodorov/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/getrandom-0.4.3/src/backends.rs	2006-07-23 21:21:28.000000000 -0400
+++ /tmp/getrandom-04-zos/src/backends.rs	2026-08-11 10:10:35.519145424 -0400
@@ -46,6 +46,7 @@
         target_os = "redox",
         target_os = "nto",
         target_os = "aix",
+        target_os = "zos",
     ))] {
         mod use_file;
         pub use use_file::*;
diff -ruN '--exclude=*.orig' '--exclude=Cargo.lock' /home/itodorov/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/getrandom-0.4.3/src/utils/get_errno.rs /tmp/getrandom-04-zos/src/utils/get_errno.rs
--- /home/itodorov/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/getrandom-0.4.3/src/utils/get_errno.rs	2006-07-23 21:21:28.000000000 -0400
+++ /tmp/getrandom-04-zos/src/utils/get_errno.rs	2026-08-11 10:42:49.110663088 -0400
@@ -1,5 +1,5 @@
 cfg_if! {
-    if #[cfg(any(target_os = "netbsd", target_os = "openbsd", target_os = "android", target_os = "cygwin"))] {
+    if #[cfg(any(target_os = "netbsd", target_os = "openbsd", target_os = "android", target_os = "cygwin", target_os = "zos"))] {
         use libc::__errno as errno_location;
     } else if #[cfg(any(target_os = "linux", target_os = "emscripten", target_os = "hurd", target_os = "redox", target_os = "dragonfly"))] {
         use libc::__errno_location as errno_location;
